pkg/agent/controller/networkpolicy/fqdn.go

```diff
@@ -0,0 +1,782 @@
+				}
+				delete(rst.ruleToSubscribers, update.ruleId)
+			}
+			// Only delete the ruleId from dirtyRules if rule realization is successful.
+			if update.err == nil {
+				rst.dirtyRules.Delete(update.ruleId)
+			}
+			rst.mutex.Unlock()
+		}
+	}
+}
+
+// notifyRuleUpdate is an interface for the reconciler to notify the ruleSyncTracker of a
+// rule realization status.
+func (f *fqdnController) notifyRuleUpdate(ruleID string, err error) {
+	f.ruleSyncTracker.updateCh <- ruleRealizationUpdate{ruleID, err}
+}
+
+func (f *fqdnController) runRuleSyncTracker(stopCh <-chan struct{}) {
+	f.ruleSyncTracker.Run(stopCh)
+}
+
+// parseDNSResponse returns the FQDN, IP query result and lowest applicable TTL of a DNS response.
+func (f *fqdnController) parseDNSResponse(msg *dns.Msg) (string, map[string]net.IP, uint32, error) {
+	if len(msg.Question) == 0 {
+		return "", nil, 0, fmt.Errorf("invalid DNS message")
+	}
+	fqdn := strings.ToLower(msg.Question[0].Name)
+	lowestTTL := uint32(math.MaxUint32) // a TTL must exist in the RRs
+	responseIPs := map[string]net.IP{}
+	for _, ans := range msg.Answer {
+		switch r := ans.(type) {
+		case *dns.A:
+			if f.ofClient.IsIPv4Enabled() {
+				responseIPs[r.A.String()] = r.A
+				if r.Header().Ttl < lowestTTL {
+					lowestTTL = r.Header().Ttl
+				}
+			}
+		case *dns.AAAA:
+			if f.ofClient.IsIPv6Enabled() {
+				responseIPs[r.AAAA.String()] = r.AAAA
+				if r.Header().Ttl < lowestTTL {
+					lowestTTL = r.Header().Ttl
+				}
+			}
+		}
+	}
+	if len(responseIPs) > 0 {
+		klog.V(4).Infof("Received DNS Packet with valid Answer. Response IPs are %v and TTL is %v", responseIPs, lowestTTL)
+	}
+	if strings.HasSuffix(fqdn, ".") {
+		fqdn = fqdn[:len(fqdn)-1]
+	}
+	return fqdn, responseIPs, lowestTTL, nil
+}
+
+func (f *fqdnController) worker() {
+	for f.processNextWorkItem() {
+	}
+}
+
+func (f *fqdnController) processNextWorkItem() bool {
+	key, quit := f.dnsQueryQueue.Get()
+	if quit {
+		return false
+	}
+	defer f.dnsQueryQueue.Done(key)
+
+	err := f.makeDNSRequest(key.(string))
+	f.handleErr(err, key)
+	return true
+}
+
+func (f *fqdnController) handleErr(err error, key interface{}) {
+	if err == nil {
+		f.dnsQueryQueue.Forget(key)
+		return
+	}
+	klog.Errorf("Error syncing fqdn name %q, retrying. Error: %v", key, err)
+	f.dnsQueryQueue.AddRateLimited(key)
+}
+
+// makeDNSRequest makes a proactive query for a FQDN to the coreDNS service.
+func (f *fqdnController) makeDNSRequest(fqdn string) error {
+	if f.dnsServerAddr == "" {
+		return fmt.Errorf("no CoreDNS server is ready for query")
+	}
+	klog.Infof("Making DNS request to coreDNS for %s", fqdn)
+	dnsClient := dns.Client{SingleInflight: true}
+	fqdnToQuery := fqdn
+	// The FQDN in the DNS request needs to end by a dot
+	if l := fqdn[len(fqdn)-1]; l != '.' {
+		fqdnToQuery = fqdn + "."
+	}
+	query := func(m *dns.Msg) (*dns.Msg, error) {
+		r, _, err := dnsClient.Exchange(m, f.dnsServerAddr)
+		if err != nil {
+			klog.ErrorS(err, "DNS exchange failed")
+		}
```

Comment:
It should return `nil, err` on error, otherwise the error checks in L645 and L655 are pointless.